### PR TITLE
search: add NOT as syntactic sugar

### LINKED ISSUE #9976: Add NOT as syntactic sugar for negatable fields
As per RFC 94, `NOT` syntax will initially serve as syntactic sugar for `-`, but later we may expand this to expressions.

### PR DESCRIPTION
**Pull request by stefanhengl**
Fixes #9976, relates to [RFC94](https://docs.google.com/document/d/1nYv_p1138eHsN7WOw0s3x69Y0HK2L8x7VP-nFKQGal8/edit)

This PR adds NOT as syntactic sugar to our query syntax. We support NOT for all negatable fields. If NOT comes before an unsupported field, we wrap the operand in a negated content field.

### Examples

|query|equivalent to|
|:---|:---|
`not file:foo`|`-file:foo `
`not repo:foo`|`-repo:foo `
`not content:foo`|`-content:foo`
`not foo`|`-content:foo `
`not foo:bar` (unknown keyword)| `-content:"foo:bar" `

### CODE REVIEW

File: internal/search/query/parser.go [hunk 1]
```diff
@@ -250,6 +267,44 @@ func (p *parser) skipSpaces() error {
 	return nil
 }
 
+// parseNegatedParameter parses `NOT field:value` or `NOT pattern` and
+// translates it to `-field:value` or `-content:pattern` respectively.
+func (p *parser) parseNegatedParameter() (Parameter, error) {
```
Comment (rvantonder) [inline internal/search/query/parser.go:272]: I think we can make a case for just calling this `parseNegatedLeafNode`, and return type `(Node, error)`. The rationale being that "Parameter" refers to the `field:value` syntax and `NOT pattern` doesn't conform to that. I realize that this function _converts_ `NOT pattern` into a `content:..` parameter ;) but we are going to later transform that back into a pattern anyway so... yeah. Then we can just return a `Pattern` node directly instead of wrapping it in a content param. But at the end of the day I'm not super attached to this idea, I just think it would simplify the indirection/transformation pipeline.

File: internal/search/query/parser.go [hunk 2]
```diff
@@ -250,6 +267,44 @@ func (p *parser) skipSpaces() error {
...
+	field, advance := ScanField(p.buf[p.pos:])
+	_, exists := allFields[strings.ToLower(field)]
+	if !exists {
+		pattern := p.ParsePattern()
+		return Parameter{
+			Field:      FieldContent,
+			Value:      pattern.Value,
+			Negated:    true,
+			Annotation: Annotation{Range: newRange(start, p.pos)},
+		}, nil
```
Comment (rvantonder) [inline internal/search/query/parser.go:291]: If you favor my suggestion above, this would simply become `return Pattern {...}`

File: internal/search/query/parser.go [hunk 3]
```diff
@@ -250,6 +267,44 @@ func (p *parser) skipSpaces() error {
 	return nil
 }
 
+// parseNegatedParameter parses `NOT field:value` or `NOT pattern` and
+// translates it to `-field:value` or `-content:pattern` respectively.
+func (p *parser) parseNegatedParameter() (Parameter, error) {
+	start := p.pos
+	p.pos += 3 // jump NOT
```
Comment (rvantonder) [inline internal/search/query/parser.go:274]:
```suggestion
	_ = p.expect(NOT) // Guaranteed to succeed.
```
The above will consume NOT.

File: internal/search/query/parser.go [hunk 4]
```diff
@@ -250,6 +267,44 @@ func (p *parser) skipSpaces() error {
...
+
+	err := p.skipSpaces()
+	if err != nil {
+		return Parameter{}, err
+	}
+
+	// if the next token is not a supported field, we parse it as -content:"..."
```
Comment (rvantonder) [inline internal/search/query/parser.go:281]: Nit: found it a bit hard to follow what the comment refers to. Possibly moving the comment to the `if` check below, or changing it to say "check whether the token is a supported field" might scan better.

Also if you favor my suggestion above, this can be rephrased: `parse it as -content:"..."` -> parse it as `NOT pattern`.

File: internal/search/query/parser_test.go [hunk 5]
```diff
@@ -60,6 +60,41 @@ func TestParseParameterList(t *testing.T) {
...
+			Want:       `{"field":"content","value":"bar","negated":true}`,
+			WantRange:  `{"start":{"line":0,"column":0},"end":{"line":0,"column":15}}`,
+			WantLabels: Regexp,
+		},
+		{
+			Name:       "NOT prefix on negated file",
+			Input:      `NOT -file:bar`,
+			Want:       `{"field":"content","value":"-file:bar","negated":true}`,
+			WantRange:  `{"start":{"line":0,"column":0},"end":{"line":0,"column":13}}`,
+			WantLabels: Regexp,
```
Comment (rvantonder) [inline internal/search/query/parser_test.go:89]: I find this test case a bit surprising :-) For a case of double negation syntax like `NOT -field` for some valid `field`, we should either:

- reject it, or
- evaluate the double negation: `NOT -field` implies `field`

I'm in favor of rejecting it, because I don't think we realistically need to support the double negation syntax. Either way, I can't convince myself that treating a syntax like `NOT -file` implies a "not" semantics that we actually run a search for. It should be easy to validate/reject the double negation syntax in the function you added. I think it's possible to just reuse `parseParameter` instead of the subparts that currently reimplement it, and then check the `ok` return value, and then inspect whether `negated` is true.

File: internal/search/query/parser_test.go [hunk 6]
```diff
@@ -874,3 +909,54 @@ func TestMergePatterns(t *testing.T) {
 		})
 	}
 }
+
+func TestMatchUnaryKeyword(t *testing.T) {
+	tests := []struct {
+		name string
+		buf  []byte
```
Comment (rvantonder) [inline internal/search/query/parser_test.go:916]: how about making this type `string` for the test cases and then converting the input to byte in `t.Run`? then we don't need the `[]byte(...)` in every test case.

File: internal/search/query/parser_test.go [hunk 7]
```diff
@@ -874,3 +909,54 @@ func TestMergePatterns(t *testing.T) {
...
+			pos:  0,
+			want: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &parser{
+				buf: tt.buf,
+				pos: tt.pos,
+			}
```
Comment (rvantonder) [inline internal/search/query/parser_test.go:956]: this can probably go on one line

File: internal/search/query/parser_test.go [hunk 8]
```diff
@@ -874,3 +909,54 @@ func TestMergePatterns(t *testing.T) {
...
+
+func TestMatchUnaryKeyword(t *testing.T) {
+	tests := []struct {
+		name string
+		buf  []byte
+		pos  int
+		want bool
+	}{
+		{
+			name: "simple NOT",
```
Comment (rvantonder) [inline internal/search/query/parser_test.go:921]: FYI, I sometimes name test cases, sometimes not, and don't feel picky about it. I find them useful when tests fail if there are a lot of similar looking inputs and it's not easy to identify the failing test case. For a short test like this I would probably `name` it, but it's up to you.

Comment (stefanhengl) [reply internal/search/query/parser_test.go:921]: With `buf` changed to type `string`, we can use  `buf` as name for `r.Run`

File: internal/search/query/parser_test.go (hunk 5 above)
Comment (stefanhengl) [reply internal/search/query/parser_test.go]: Yeah, I wasn't sure about this one either. I agree with you to just reject it and let the user know how to fix it.
